modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

```diff
@@ -1040,27 +1064,39 @@ private[smithy4s] class SchemaVisitorJCodec(
   )(dispatch: Alt.Dispatcher[U]): JCodec[U] =
     new TaggedUnionJCodec[U](alternatives)(dispatch) {
 
-      def decodeValue(cursor: Cursor, in: JsonReader): U =
+      def decodeValue(cursor: Cursor, in: JsonReader): U = {
+        // ! - I'm not sure about this
+        in.setMark()
         if (in.isNextToken('{')) {
           if (in.isNextToken('}'))
             in.decodeError("Expected a single key/value pair")
           else {
             in.rollbackToken()
             val key = in.readKeyAsString()
-            cursor.push(key)
             val handler = handlerMap.get(key)
-            if (handler eq null) in.discriminatorValueError(key)
-            val result = handler(cursor, in)
-            cursor.pop()
-            if (in.isNextToken('}')) result
-            else {
-              in.rollbackToken()
-              in.decodeError(s"Expected no other field after $key")
+            if ((handler eq null) && (unknownTagHandler eq null))
+              in.discriminatorValueError(key)
+
+            if (handler ne null) {
+              cursor.push(key)
+              val result = handler(cursor, in)
+              cursor.pop()
+              if (in.isNextToken('}')) result
+              else {
+                in.rollbackToken()
+                in.decodeError(s"Expected no other field after $key")
+              }
+            } else {
+              // ! - I'm not sure about this
+              in.rollbackToMark()
+              unknownTagHandler(cursor, in)
```

Comment:
I think you need to check for non-nullity of `unknownTagHandler` here

Reply:
not necessarily, if it's null then `in.discriminatorValueError(key)` kicks in, in the previous conditional

Reply:
I'll rewrite these two ifs to just one, if that helps readability - will compare

Reply:
turned into a single `if` with 3 branches 65e916b1d3aeca5cef56c5f603dba48961656cf0